File: r/sparklabs/batch/sales/dataframe/product_rank.R
source("r/sparklabs/batch/sales/dataframe/utils.R")

order_counts <- agg(groupBy(order_items, order_items$product_id), order_count = count(order_items$product_id))
category_counts <- agg(
  groupBy(join(order_counts, products, order_counts$product_id == products$product_id), products$category),
  category_count = count(products$category)
)
result <- withColumn(
  category_counts,
  "category_rank",
  dense_rank() %over% orderBy(windowPartitionBy(), desc("category_count"))
)
showDF(result, truncate = FALSE)

File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/dataframe/ProductRank.scala
package io.devorbit.sparklabs.batch.sales.dataframe

import org.apache.spark.sql.expressions.Window
import org.apache.spark.sql.functions._

object ProductRank extends App {

  val windowSpec = Window
    .orderBy(desc("category_count"))

  Utils.orderItems
    .groupBy(col("product_id"))
    .agg(count("*").as("order_count"))
    .join(Utils.products, Seq("product_id"), "inner")
    .groupBy(col("category"))
    .agg(count("*").as("category_count"))
    .withColumn("category_rank", dense_rank().over(windowSpec))
    .show(false)

//  Thread.sleep(1000000)
}
